In [5]:
!pip install youtube-transcript-api

In [6]:
from youtube_transcript_api import YouTubeTranscriptApi

In [18]:
from youtube_transcript_api import YouTubeTranscriptApi
import re
from collections import Counter

In [14]:
def get_youtube_video_id(url):
    """
    Extract the video ID from a YouTube URL.
    """
    regex = r'(?:https?://)?(?:www\.)?(?:youtube\.com|youtu\.be)/(?:watch\?v=)?([a-zA-Z0-9_-]+)'
    match = re.match(regex, url)
    if match:
        return match.group(1)
    return None

def get_transcript(video_id):
    """
    Fetch the transcript for the given YouTube video ID.
    """
    try:
        transcript_list = YouTubeTranscriptApi.list_transcripts(video_id)
        transcript = transcript_list.find_transcript(['en'])
        return {'transcript': transcript.fetch()}
    except Exception as e:
        return str(e)

def main(url):
    video_id = get_youtube_video_id(url)

    if not video_id:
        print("Invalid YouTube URL")
        return

    transcript_data = get_transcript(video_id)

    if isinstance(transcript_data, str):
        print(f"Error: {transcript_data}")
    else:
        transcript = transcript_data['transcript']
        transcript_text = " ".join([item['text'] for item in transcript])
        return transcript_text


In [15]:
url='https://youtu.be/QUTYxwTsbiM?si=P4NmtNI53q0Bli7S'

In [19]:
# Get the transcript text
transcript_text = main(youtube_url)
if transcript_text:
    print(transcript_text)

      The name of the story is Clever Fish One day, a fisher man would fishing to a river as usual He throw his net into the river And he just sat waiting there for fish together So that He could sell a lot of fish in the market and get some good money from the fish after sometimes fisher man hooked relevant basil in the net Thinking that, he must got a lot of fish in the net He acted to count the net out of water But then   he saw there is one tiny little fish in that net   but then suddenly the fish start to taking to him said to the fisher man oh fisher man, please please please, leave me please leave me   but fisher man did not give any attention to the request of the fish But then again, the tiny little fish said to the fisher man oh fisher man, I will tell you something which is a huge help If you leave me back in water I will tell all my friends about you. And I bring them to the bank coffee first. So that, when you come next time you have much more fish Fisher man talk to himse

In [20]:
!pip install youtube-transcript-api textblob


In [21]:
from textblob import TextBlob


In [24]:
# Function to normalize text (basic text normalization)
def normalize_text(transcript_text):
    # Convert to lowercase
    transcript_text = transcript_text.lower()
    # Remove special characters and numbers
    transcript_text = re.sub(r'[^a-z\s]', '', text)
    return transcript_text

# Function for spelling correction using TextBlob
def correct_spelling(transcript_text):
    blob = TextBlob(transcript_text)
    corrected_text = str(blob.correct())
    return corrected_text

In [26]:
print(correct_spelling(transcript_text))

      The name of the story is Clever Wish One day, a fishes man would fishing to a river as usual He throw his net into the river And he just sat waiting there for fish together To that He could sell a lot of fish in the market and get some good money from the fish after sometimes fishes man hooked relevant basil in the net Thinking that, he must got a lot of fish in the net He acted to count the net out of water But then   he saw there is one tiny little fish in that net   but then suddenly the fish start to taking to him said to the fishes man oh fishes man, please please please, leave me please leave me   but fishes man did not give any attention to the request of the fish But then again, the tiny little fish said to the fishes man oh fishes man, I will tell you something which is a huge help Of you leave me back in water I will tell all my friends about you. And I bring them to the bank coffee first. To that, when you come next time you have much more fish Either man talk to himse

In [28]:
!pip install rake_nltk
from rake_nltk import Rake
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [29]:
def extract_keywords_rake(transcript_texttext, top_n=10):
    rake = Rake()  # Rake object from the rake-nltk package
    rake.extract_keywords_from_text(transcript_text)
    keywords = rake.get_ranked_phrases()[:top_n]
    return keywords

In [32]:
from nltk.corpus import stopwords
nltk.download('stopwords')


[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


True

In [33]:
print("Transcript:\n", transcript_text[:500], "...\n")
print("Top Keywords using RAKE:\n", extract_keywords_rake(transcript_text))

Transcript:
       The name of the story is Clever Fish One day, a fisher man would fishing to a river as usual He throw his net into the river And he just sat waiting there for fish together So that He could sell a lot of fish in the market and get some good money from the fish after sometimes fisher man hooked relevant basil in the net Thinking that, he must got a lot of fish in the net He acted to count the net out of water But then   he saw there is one tiny little fish in that net   but then suddenly th ...

Top Keywords using RAKE:
 ['found ... one tiny little fish today tomorrow', 'sometimes fisher man hooked relevant basil', 'fisher man oh fisher man', 'fisher man oh fisher man', 'come back poor fisher man', 'one tiny little fish', 'fisher man would fishing', 'fish fisher man talk', 'tiny little fish said', 'clever fish one day']


In [36]:
!pip install youtube-transcript-api nltk transformers pyttsx3
from nltk import sent_tokenize
from transformers import pipeline
import pyttsx3

In [44]:
# Extractive Summarization using TextRank (NLTK)
def extractive_summarization(transcript_text, num_sentences=5):
    sentences = sent_tokenize(transcript_text)
    if len(sentences) > num_sentences:
        summary = ' '.join(sentences[:num_sentences])
    else:
        summary = transcript_text
    return summary


In [45]:
# Abstractive Summarization using BART (transformers)
def abstractive_summarization(transcript_text, min_length=30, max_length=100):
    summarizer = pipeline('summarization', model='facebook/bart-large-cnn')
    summary = summarizer(transcript_text, min_length=min_length, max_length=max_length)[0]['summary_text']
    return summary

# Text-to-Speech for Read-Aloud Feature
def read_aloud(transcript_text):
    engine = pyttsx3.init()
    engine.say(transcript_text)
    engine.runAndWait()

In [47]:
print("Transcript:\n", transcript_text[:500], "...\n")
 # Extractive Summarization
extractive_summary = extractive_summarization(transcript_text)
print("Extractive Summary:\n", extractive_summary)



# Abstractive Summarization
abstractive_summary = abstractive_summarization(transcript_text)
print("Abstractive Summary:\n", abstractive_summary)



Transcript:
       The name of the story is Clever Fish One day, a fisher man would fishing to a river as usual He throw his net into the river And he just sat waiting there for fish together So that He could sell a lot of fish in the market and get some good money from the fish after sometimes fisher man hooked relevant basil in the net Thinking that, he must got a lot of fish in the net He acted to count the net out of water But then   he saw there is one tiny little fish in that net   but then suddenly th ...

Extractive Summary:
       The name of the story is Clever Fish One day, a fisher man would fishing to a river as usual He throw his net into the river And he just sat waiting there for fish together So that He could sell a lot of fish in the market and get some good money from the fish after sometimes fisher man hooked relevant basil in the net Thinking that, he must got a lot of fish in the net He acted to count the net out of water But then   he saw there is one tiny little

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:89: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/1.58k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/1.63G [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/363 [00:00<?, ?B/s]

vocab.json:   0%|          | 0.00/899k [00:00<?, ?B/s]

merges.txt:   0%|          | 0.00/456k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.36M [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1601: FutureWarning: `clean_up_tokenization_spaces` was not set. It will be set to `True` by default. This behavior will be depracted in transformers v4.45, and will be then set to `False` by default. For more details check this issue: https://github.com/huggingface/transformers/issues/31884
  warnings.warn(


Abstractive Summary:
 Fisher man would fishing to a river as usual He throw his net into the river and he just sat waiting there for fish together. He could sell a lot of fish in the market and get some good money from the fish after sometimes fisher man hooked relevant basil in the net. One day he saw there is one tiny little fish in that net   but then suddenly the fish start to taking to him.


In [ ]:
# Read aloud the extractive summary
print("\nReading Extractive Summary Aloud:")
read_aloud(extractive_summary)

# Read aloud the abstractive summary
print("\nReading Abstractive Summary Aloud:")
read_aloud(abstractive_summary)

In [49]:
# Sentiment analysis
def analyze_sentiment(transcript_texttext):
    # Create a TextBlob object and get sentiment polarity and subjectivity
    blob = TextBlob(transcript_text)
    sentiment_polarity = blob.sentiment.polarity
    sentiment_subjectivity = blob.sentiment.subjectivity
    return sentiment_polarity, sentiment_subjectivity

In [50]:
print("Transcript Fetched Successfully!")
print("Performing Sentiment Analysis...")

# Analyze sentiment
polarity, subjectivity = analyze_sentiment(transcript_text)

# Display the results
print(f"Sentiment Polarity: {polarity} (Range: -1 to 1)")
print(f"Sentiment Subjectivity: {subjectivity} (Range: 0 to 1)")

Transcript Fetched Successfully!
Performing Sentiment Analysis...
Sentiment Polarity: 0.10076754385964912 (Range: -1 to 1)
Sentiment Subjectivity: 0.4546052631578948 (Range: 0 to 1)


The sentiment analysis shows that the transcript has a slightly positive tone, with a polarity of 0.1008, indicating mild positivity. The subjectivity score of 0.4546 suggests a balance between objective information and subjective opinions, meaning the content contains both facts and some personal or emotional expressions.